transform_aes.c

```diff
@@ -219,28 +220,37 @@ static int transop_encode_aes( n2n_trans_op_t * arg,
             /* Encode the security association (SA) number */
             encode_uint32( outbuf, &idx, sa->sa_id );
 
-            /* Encrypt the assembly contents and write the ciphertext after the SA. */
-            len = in_len + TRANSOP_AES_NONCE_SIZE;
+           /* The following fields (message authentication, IV) are calculated respectively filled after data encrpytion */
 
-            /* The assembly buffer is a source for encrypting data. The nonce is
+            /* Encrypt the assembly contents and write the ciphertext after the IV. */
+            len = in_len + TRANSOP_AES_DATA_LEN;
+
+            /* The assembly buffer is a source for encrypting data. The data length is
              * written in first followed by the packet payload. The whole
              * contents of assembly are encrypted. */
-            pnonce = (uint32_t *)assembly;
-            *pnonce = rand();
-            memcpy( assembly + TRANSOP_AES_NONCE_SIZE, inbuf, in_len );
+            data_len = (uint16_t *)assembly;
+            *data_len = in_len;
```

Comment:
In order to make connection between big endian and little endian devices work properly, this must be encoded in network byte order. Check out https://en.wikipedia.org/wiki/Endianness for details. Please replace `*data_len = in_len;` with `*data_len = htons(in_len);`. By the way, why sending payload length required here?

Reply:
Thank you, I surely will change that.

The data length is required here in the first place because we need to keep the encrypted data "prefix free", meaning that no other ever encrypted message could be put in front of any other package encrypted data forming a valid packet again. To put packet's length upfront is an effective countermeasure to prevent a variable-length attack described [here](https://en.wikipedia.org/wiki/CBC-MAC).